File: Algorithms/149.MaxPointsOnALine.cpp
class Solution {
public:
    int gcd(int a, int b) {
        return (b == 0) ? a : gcd(b, a % b);
    }
    int maxPoints(vector<Point>& points) {
        int size = points.size();
        if (size <= 2)
            return size;
        map<pair<int, int>, int> slopeMap;
        int maxCount = 0;
        for (int i=0; i < size; ++i) {
            int duplicates = 1;
            int vertical = 0;
            slopeMap.clear();
            for (int j=0; j < size; ++j) {
                if (i == j)
                    continue;
                if (points[i].x == points[j].x && points[i].y == points[j].y) {
                    duplicates ++;
                    continue;
                }
                if (points[i].x == points[j].x) {
                    vertical ++;
                    continue;
                }
                int dy = points[i].y - points[j].y;
                int dx = points[i].x - points[j].x;
                int d = gcd(dx, dy);
                slopeMap[{dx / d, dy / d}] ++;
            }
            for (auto it = slopeMap.begin(); it != slopeMap.end(); ++it) {
                if (it->second > vertical)
                    vertical = it->second;
            }
            maxCount = max(vertical + duplicates, maxCount);
        }
        return maxCount;
    }
};

File: Algorithms/py/149.MaxPointsOnALine.py
class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[Point]
        :rtype: int
        """
        def gcd(a, b):
            if b == 0:
                return a
            return gcd(b, a % b)
        size = len(points)
        if size <= 2:
            return size
        slopeMap = collections.defaultdict(int)
        maxCount = 0
        for i in range(size):
            duplicates, vertical = 1, 0
            slopeMap.clear()
            for j in range(size):
                if i == j:
                    continue
                if points[i].x == points[j].x and points[i].y == points[j].y:
                    duplicates += 1
                    continue
                if points[i].x == points[j].x:
                    vertical += 1
                    continue
                dy, dx = points[i].y - points[j].y, points[i].x - points[j].x
                d = gcd(dy, dx)
                slopeMap[dx / d, dy / d] += 1
            if slopeMap:
                vertical = max(vertical, max(slopeMap.values()))
            maxCount = max(vertical + duplicates, maxCount)
        return maxCount
